### PR TITLE
Make `invariant` fully type safe

### PR DESCRIPTION
**Pull request by desfero**
**Technical notes**
Extracted from #4226 to keep PR clean from web app changes.

### CODE REVIEW

File: packages/shared/src/utils/invariant.ts [hunk 1]
```diff
@@ -1,12 +1,17 @@
...
-  type = "InvariantError";
+  constructor(message: string) {
+    super(`InvariantError: ${message}`);
+  }
 }
 
-export function invariant(value: any, message: string): null {
-  if (!value) {
+/**
+ * Assets that the given condition is truthy
```
Comment (desfero) [inline packages/shared/src/utils/invariant.ts:8]:
```suggestion
 * Asserts that the given condition is truthy
```

File: packages/shared/src/utils/invariant.ts [hunk 2]
```diff
@@ -1,12 +1,17 @@
...
+ *
+ * @param condition - Either truthy or falsy value
+ * @param message - An error message is required so it's easier to figure out the error
```
Comment (desfero) [inline packages/shared/src/utils/invariant.ts:11]:
```suggestion
 * @param message - An error message
```

File: packages/shared/src/utils/invariant.ts [hunk 3]
```diff
@@ -1,12 +1,17 @@
...
+ */
+export function invariant(condition: any, message: string): asserts condition {
+  if (!condition) {
     throw new InvariantError(message);
   }
-
-  // Useful in react component context as `undefined` is not allowed in JSX
-  return null;
```
Comment (desfero) [inline packages/shared/src/utils/invariant.ts:11]: not possible to have return type other than `never | void`

File: packages/shared/src/utils/invariant.spec.ts [hunk 4]
```diff
@@ -13,4 +14,12 @@ describe("invariant", () => {
     invariant(true, "description");
     invariant(1, "description");
   });
+
+  it("should properly narrow type", () => {
```
Comment (arjunu) [inline packages/shared/src/utils/invariant.spec.ts:18]: why not put this in `./test/testUtils.spec.ts` ?

**Pull request merged**